jupyter_client/channels.py

```diff
@@ -34,24 +38,27 @@ class HBChannel(Thread):
     this channel, the kernel manager will ensure that it is paused and un-paused
     as appropriate.
     """
-    context = None
     session = None
     socket = None
     address = None
     _exiting = False
 
-    time_to_dead = 1.
-    poller = None
+    time_to_dead: float = 1.
     _running = None
     _pause = None
     _beating = None
 
-    def __init__(self, context=None, session=None, address=None, loop=None):
+    def __init__(
+        self,
+        context: zmq.asyncio.Context,
```

Comment:
It looks like the removal of the `None` default is what is breaking `qtconsole`, although this does imply they must not be using the HB channel since it unconditionally references through `self.context` in its `_create_socket()` method.

Reply:
> It looks like the removal of the None default is what is breaking qtconsole

Thanks for your help @kevin-bates. I found that too but now we have another problem (see below).

> although this does imply they must not be using the HB channel since it unconditionally references through self.context in its _create_socket() method.

It has something to do with the way the channel inherits from HBChannel and QObject and a re-implementation of `super` to make that work (I really don't understand that logic, @minrk implemented it).

Reply:
This is the error I'm seeing now while starting a kernel:

```python
ERROR:tornado.general:Uncaught exception in ZMQStream callback
Traceback (most recent call last):
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 434, in _run_callback
    callback(*args, **kwargs)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/threaded.py", line 101, in _handle_recv
    ident, smsg = self.session.feed_identities(msg)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/session.py", line 954, in feed_identities
    idx = msg_list.index(DELIM)
AttributeError: '_asyncio.Future' object has no attribute 'index'
ERROR:tornado.general:Uncaught exception in zmqstream callback
Traceback (most recent call last):
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 451, in _handle_events
    self._handle_recv()
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 480, in _handle_recv
    self._run_callback(callback, msg)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 434, in _run_callback
    callback(*args, **kwargs)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/threaded.py", line 101, in _handle_recv
    ident, smsg = self.session.feed_identities(msg)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/session.py", line 954, in feed_identities
    idx = msg_list.index(DELIM)
AttributeError: '_asyncio.Future' object has no attribute 'index'
ERROR:asyncio:Exception in callback BaseAsyncIOLoop._handle_events(30, 1)
handle: <Handle BaseAsyncIOLoop._handle_events(30, 1)>
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/tornado/platform/asyncio.py", line 189, in _handle_events
    handler_func(fileobj, events)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 451, in _handle_events
    self._handle_recv()
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 480, in _handle_recv
    self._run_callback(callback, msg)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/zmq/eventloop/zmqstream.py", line 434, in _run_callback
    callback(*args, **kwargs)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/threaded.py", line 101, in _handle_recv
    ident, smsg = self.session.feed_identities(msg)
  File "/home/carlos/.virtualenvs/test-jupyter-client/lib/python3.8/site-packages/jupyter_client/session.py", line 954, in feed_identities
    idx = msg_list.index(DELIM)
AttributeError: '_asyncio.Future' object has no attribute 'index'
```

Any ideas?

Reply:
I also saw this @ccordoba12, it has to do with the threaded channels. Don't worry, I will clear all that out before the next release.

Reply:
Great! Please give me a ping in case you want me to test your changes.